### PR TITLE
ensure that integers cast to pointers will never point at a valid alloc, not even the zst alloc

### LINKED ISSUE #77: triggering `assume` in a small example that rustc compiles fine
repro example:

```rust
fn main() {
    vec![()].into_iter();
}
```

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #77

### CODE REVIEW

**Comment by solson** [comment]
Sorry for the delay. I'll try to do a full review this evening.

**Comment by oli-obk** [comment]
No hurry. I'm not gonna get to it before monday

File: src/interpreter/mod.rs [hunk 1]
```diff
@@ -523,7 +519,8 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
                                 self.write_value(value, dest, value_ty)?;
                             } else {
                                 assert_eq!(operands.len(), 0);
-                                let zero = self.isize_primval(0);
+                                let value_size = self.type_size(dest_ty).expect("pointer types are sized");
+                                let zero = PrimVal::from_int_with_size(0, value_size);
```
Comment (solson) [inline src/interpreter/mod.rs]: Why this change? I think we can assume `isize` is pointer sized, and `isize_primval` does mean the target `isize`.

File: src/interpreter/mod.rs [hunk 2]
```diff
@@ -541,13 +538,14 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
                                 }
-                                let offset = self.nonnull_offset(dest_ty, nndiscr, discrfield)?;
+                                let (offset, ty) = self.nonnull_offset_and_ty(dest_ty, nndiscr, discrfield)?;
 
                                 // FIXME(solson)
                                 let dest = self.force_allocation(dest)?.to_ptr();
 
                                 let dest = dest.offset(offset.bytes() as isize);
-                                try!(self.memory.write_isize(dest, 0));
+                                let dest_size = self.type_size(ty).unwrap_or(self.memory.pointer_size());
```
Comment (solson) [inline src/interpreter/mod.rs]: In what cases can this default to `self.memory.pointer_size()`?

File: src/interpreter/terminator/mod.rs [hunk 3]
```diff
@@ -263,14 +263,17 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
+                self.read_nonnull_discriminant_value(adt_ptr, nndiscr, discr_size)?
             }
 
             StructWrappedNullablePointer { nndiscr, ref discrfield, .. } => {
-                let offset = self.nonnull_offset(adt_ty, nndiscr, discrfield)?;
+                let (offset, ty) = self.nonnull_offset_and_ty(adt_ty, nndiscr, discrfield)?;
                 let nonnull = adt_ptr.offset(offset.bytes() as isize);
-                self.read_nonnull_discriminant_value(nonnull, nndiscr)?
+                // only the pointer part of a fat pointer is used for this space optimization
+                let discr_size = self.type_size(ty).unwrap_or(self.memory.pointer_size());
```
Comment (solson) [inline src/interpreter/terminator/mod.rs]: Ditto here. When can the nonnull field be a unsized?

File: src/interpreter/mod.rs (hunk 1 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: We don't want an isize here. The rawnullablepointer opt also happens for nonpointers

File: src/interpreter/terminator/mod.rs (hunk 3 above)
Comment (oli-obk) [inline src/interpreter/terminator/mod.rs]: It can't be, but the MIR ends up giving us an index path that will give us the first field of the fat pointer, which ends up being te unsized value. Not sure if the code giving us the type path is wrong or if this is intended. I'll check and get back to you

Comment (eddyb) [inline src/interpreter/terminator/mod.rs]: The first field of the fat pointer is the thin data pointer. This is a GEP field path, it will not deref nested pointers, just offset into a flat structure.

File: src/interpreter/mod.rs (hunk 1 above)
Comment (solson) [inline src/interpreter/mod.rs]: Ahh, right, for NonZero things that aren't pointer-sized? I hate the `RawNullablePointer` names in layout...

File: src/interpreter/mod.rs (hunk 2 above)
Comment (solson) [inline src/interpreter/mod.rs]: What unsized case? ty is the type of the non-zero field that is used for the discriminant in this optimization, right? How can that field be unsized?

File: src/interpreter/terminator/mod.rs (hunk 3 above)
Comment (solson) [inline src/interpreter/terminator/mod.rs]: Ah, I see. It sounds like we'd need to adjust the type, then, to match @eddyb's point. Alternately, maybe it can return the size instead of the type?

Comment (oli-obk) [inline src/interpreter/terminator/mod.rs]: fixed

File: src/interpreter/mod.rs [hunk 4]
```diff
@@ -734,8 +734,17 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                assert_eq!(field_index, 0);
-                Ok(ty)
+                if self.type_is_sized(ty) {
+                    assert_eq!(field_index, 0);
+                    Ok(self.tcx.mk_imm_ptr(self.tcx.types.never))
+                } else {
+                    match (field_index, &ty.sty) {
+                        (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
+                        (1, &ty::TyTrait(_)) |
+                        (0, _) => Ok(self.tcx.mk_imm_ptr(self.tcx.types.never)),
```
Comment (eddyb) [inline src/interpreter/mod.rs]: I'd use a pointer to `u8` to avoid any other implications. Btw you need to use `struct_tail` to get `ty`.
Also I don't think the sized case above makes sense? We only treat a fat pointer like a pair.

File: src/interpreter/mod.rs [hunk 5]
```diff
@@ -734,16 +734,12 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                } else {
-                    match (field_index, &ty.sty) {
-                        (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
-                        (1, &ty::TyTrait(_)) |
-                        (0, _) => Ok(self.tcx.mk_imm_ptr(self.tcx.types.never)),
-                        _ => bug!("invalid fat pointee type"),
-                    }
+                assert!(!self.type_is_sized(ty));
+                match (field_index, &self.tcx.struct_tail(ty).sty) {
+                    (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
```
Comment (eddyb) [inline src/interpreter/mod.rs]: This should also mention `TyStr`... *However*, neither `[T]` nor `str`'s length fields can be used for the optimization.
It's probably okay to keep it? I'm not sure. Also, the `assert!(!self.type_is_sized(ty));` can be removed since you use `struct_tail`, all you have to do is match on `TyTrait` / `TySlice` / `TyStr` specifically for the pointer field.

**Comment by oli-obk** [comment]
we actually pass a lot more tests now: https://github.com/solson/miri/issues/55

File: src/interpreter/mod.rs (hunk 5 above)
Comment (solson) [inline src/interpreter/mod.rs]: > It's probably okay to keep it?

I'd prefer to stick to cases we know can happen (`field_index == 0`, or more specifically the constant that says which fat pointer field is the thin pointer) and assert that nothing else happens. Since we can't use length fields, I'd remove any code that handles those cases. Vtable pointers seem plausible, but do we even use those?

File: src/interpreter/mod.rs [hunk 6]
```diff
@@ -734,16 +734,12 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                    Ok(self.tcx.mk_imm_ptr(self.tcx.types.never))
-                } else {
-                    match (field_index, &ty.sty) {
-                        (1, &ty::TySlice(_)) => Ok(self.tcx.types.usize),
-                        (1, &ty::TyTrait(_)) |
-                        (0, _) => Ok(self.tcx.mk_imm_ptr(self.tcx.types.never)),
-                        _ => bug!("invalid fat pointee type"),
-                    }
+                assert!(!self.type_is_sized(ty));
+                match (field_index, &self.tcx.struct_tail(ty).sty) {
```
Comment (solson) [inline src/interpreter/mod.rs]: I think the `&` here and on the patterns is redundant.

File: src/interpreter/mod.rs (hunk 5 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: What about other values used as discriminants? The slice length might be used in optimizations for nonzst slices, which can't be usize::max_value() long.

Comment (solson) [inline src/interpreter/mod.rs]: Actually, I take back what I said since this is the general `get_field_ty`, not specific to the non-null optimizations.

Comment (solson) [inline src/interpreter/mod.rs]: The fact that this is the general `get_field_ty` also means we should guard against thin pointer types one way or another, perhaps by adding a match guard so thin pointers go to the "can't handle type" case.

File: src/interpreter/mod.rs (hunk 6 above)
Comment (solson) [inline src/interpreter/mod.rs]: @eddyb Is this a problem because of the tuple? Not being copyable for sure wouldn't be a problem when matching directly, because the patterns don't try to copy (or move, or bind at all) anything in the `TyS`.

Comment (solson) [inline src/interpreter/mod.rs]: In other words, the error is on building the tuple, not matching... you'd need language support for matching two things at once without a tuple to make it work the way I expected.

File: src/interpreter/mod.rs (hunk 5 above)
Comment (oli-obk) [inline src/interpreter/mod.rs]: The sized assertion does this check

File: src/error.rs [hunk 7]
```diff
@@ -61,7 +61,7 @@ impl<'tcx> Error for EvalError<'tcx> {
             EvalError::DanglingPointerDeref =>
                 "dangling pointer was dereferenced",
             EvalError::InvalidFunctionPointer =>
-                "tried to use a pointer as a function pointer",
+                "tried to use an integer pointer as a function pointer",
```
Comment (oli-obk) [inline src/error.rs]: Nothing changed here, the error message was outdated way before this pr, we can adjust the variant name to match that

Comment (solson) [inline src/error.rs]: What kind of error do we actually get when treating a regular pointer as a function pointer?

**Comment by oli-obk** [comment]
ok, everything has been addressed and fixed. I did some changes to how Primval pointers are handled to make sure that there's no difference between reading a primval from memory and transmuting a primval

File: src/interpreter/mod.rs [hunk 8]
```diff
@@ -963,33 +963,27 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                    }
+                trace!("deref to {} on {:?}", pointee_type, val);
 
-                    ByVal(ptr) if ptr.try_as_ptr().is_some() => {
-                        let ptr = ptr.try_as_ptr().unwrap();
-                        (ptr, LvalueExtra::None)
-                    }
-
-                    _ => bug!("can't deref non pointer types"),
+                match self.tcx.struct_tail(pointee_type).sty {
```
Comment (solson) [inline src/interpreter/mod.rs]: This is so much clearer. :D

File: src/memory.rs [hunk 9]
```diff
@@ -533,8 +533,8 @@ impl<'a, 'tcx> Memory<'a, 'tcx> {
     }
 
     pub fn write_primval(&mut self, dest: Pointer, val: PrimVal) -> EvalResult<'tcx, ()> {
-        if let Some(ptr) = val.try_as_ptr() {
-            return self.write_ptr(dest, ptr);
+        if let Some(alloc_id) = val.relocation {
+            return self.write_ptr(dest, Pointer::new(alloc_id, val.bits as usize));
```
Comment (solson) [inline src/memory.rs]: Not your mistake, just making a note for posterity: this `as usize` is wrong when emulating a 64-bit target on a 32-bit host. I have a TODO on `Pointer` for changing the `usize` to `u64`.

**Comment by solson** [comment]
The trend toward making decisions in the code based on `Ty`s rather than `PrimValKind` or other Miri things like relocations is a great trend. It will help with some of my refactorings and seems to be making the code clearer.
